### ISSUE TITLE
[BUG] Improperly matched output dtypes detected in col, float64 and int64

### ISSUE THREAD

**Issue by rnyak** [NVIDIA-Merlin/NVTabular#1396]
**Describe the bug**
I have this pipeline with a customer dataset and I am getting the following error. The ` c_156` dtype is originally `int64`. In the schema it appears as `FLOAT`. I dont know why. 

```
def reduce_mem(col):
    if np.issubdtype(col.dtype, np.integer):
        if (col.min()>-32768)&(col.max()<32768):
            col = col.astype('int16')
        elif (col.min()>-2147483648)&(col.max()<2147483648):
            col = col.astype('int32')
    if np.issubdtype(col.dtype, np.float):
        col = col.astype('float32')
    return col

reduced_cols = cols >> nvt.ops.LambdaOp(reduce_mem)

dataset = nvt.Dataset(data_path, part_size='500MB')
workflow = nvt.Workflow(reduced_cols)
workflow.fit(dataset)
# Write to new "shuffled" and "processed" dataset
workflow.transform(dataset).to_parquet(
    OUTPUT_FOLDER,
    out_files_per_proc=1,
    shuffle=False
)
```
```
Failed to transform operator <nvtabular.graph.ops.selection.SelectionOp object at 0x7f5504ebd190>
Traceback (most recent call last):
  File "/nvtabular/nvtabular/workflow/workflow.py", line 485, in _transform_partition
    raise TypeError(
TypeError: Improperly matched output dtypes detected in c_156, float64 and int64
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File <timed exec>:44, in <module>

File /nvtabular/nvtabular/io/dataset.py:901, in Dataset.to_parquet(self, output_path, shuffle, preserve_files, output_files, out_files_per_proc, num_threads, dtypes, cats, conts, labels, suffix, partition_on, method, write_hugectr_keyset)
    898 self.schema.write(output_path)
    900 # Output dask_cudf DataFrame to dataset
--> 901 _ddf_to_dataset(
    902     ddf,
    903     fs,
    904     output_path,
    905     shuffle,
    906     output_files,
    907     out_files_per_proc,
    908     cats or [],
    909     conts or [],
    910     labels or [],
    911     "parquet",
    912     num_threads,
    913     self.cpu,
    914     suffix=suffix,
    915     partition_on=partition_on,
    916     schema=self.schema if write_hugectr_keyset else None,
    917 )

File /nvtabular/nvtabular/io/dask.py:370, in _ddf_to_dataset(ddf, fs, output_path, shuffle, file_partition_map, out_files_per_proc, cat_names, cont_names, label_names, output_format, num_threads, cpu, suffix, partition_on, schema)
    368     out = client.compute(out).result()
    369 else:
--> 370     out = dask.compute(out, scheduler="synchronous")[0]
    372 if cached_writers:
    373     # Follow-up Shuffling and _metadata creation
    374     _finish_dataset(client, ddf, output_path, fs, output_format, cpu, schema)

File /usr/local/lib/python3.8/dist-packages/dask/base.py:571, in compute(traverse, optimize_graph, scheduler, get, *args, **kwargs)
    568     keys.append(x.__dask_keys__())
    569     postcomputes.append(x.__dask_postcompute__())
--> 571 results = schedule(dsk, keys, **kwargs)
    572 return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

File /usr/local/lib/python3.8/dist-packages/dask/local.py:553, in get_sync(dsk, keys, **kwargs)
    548 """A naive synchronous version of get_async
    549 
    550 Can be useful for debugging.
    551 """
    552 kwargs.pop("num_workers", None)  # if num_workers present, remove it
--> 553 return get_async(
    554     synchronous_executor.submit,
    555     synchronous_executor._max_workers,
    556     dsk,
    557     keys,
    558     **kwargs,
    559 )

File /usr/local/lib/python3.8/dist-packages/dask/local.py:496, in get_async(submit, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, chunksize, **kwargs)
    494 while state["waiting"] or state["ready"] or state["running"]:
    495     fire_tasks(chunksize)
--> 496     for key, res_info, failed in queue_get(queue).result():
    497         if failed:
    498             exc, tb = loads(res_info)

File /usr/lib/python3.8/concurrent/futures/_base.py:437, in Future.result(self, timeout)
    435     raise CancelledError()
    436 elif self._state == FINISHED:
--> 437     return self.__get_result()
    439 self._condition.wait(timeout)
    441 if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:

File /usr/lib/python3.8/concurrent/futures/_base.py:389, in Future.__get_result(self)
    387 if self._exception:
    388     try:
--> 389         raise self._exception
    390     finally:
    391         # Break a reference cycle with the exception in self._exception
    392         self = None

File /usr/local/lib/python3.8/dist-packages/dask/local.py:538, in SynchronousExecutor.submit(self, fn, *args, **kwargs)
    536 fut = Future()
    537 try:
--> 538     fut.set_result(fn(*args, **kwargs))
    539 except BaseException as e:
    540     fut.set_exception(e)

File /usr/local/lib/python3.8/dist-packages/dask/local.py:234, in batch_execute_tasks(it)
    230 def batch_execute_tasks(it):
    231     """
    232     Batch computing of multiple tasks with `execute_task`
    233     """
--> 234     return [execute_task(*a) for a in it]

File /usr/local/lib/python3.8/dist-packages/dask/local.py:234, in <listcomp>(.0)
    230 def batch_execute_tasks(it):
    231     """
    232     Batch computing of multiple tasks with `execute_task`
    233     """
--> 234     return [execute_task(*a) for a in it]

File /usr/local/lib/python3.8/dist-packages/dask/local.py:225, in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    223     failed = False
    224 except BaseException as e:
--> 225     result = pack_exception(e, dumps)
    226     failed = True
    227 return key, result, failed

File /usr/local/lib/python3.8/dist-packages/dask/local.py:220, in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    218 try:
    219     task, data = loads(task_info)
--> 220     result = _execute_task(task, data)
    221     id = get_id()
    222     result = dumps((result, id))

File /usr/local/lib/python3.8/dist-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File /usr/lib/python3.8/contextlib.py:75, in ContextDecorator.__call__.<locals>.inner(*args, **kwds)
     72 @wraps(func)
     73 def inner(*args, **kwds):
     74     with self._recreate_cm():
---> 75         return func(*args, **kwds)

File /nvtabular/nvtabular/io/dask.py:203, in _write_subgraph(subgraph, fns, output_path, shuffle, fs, cat_names, cont_names, label_names, output_format, num_threads, cpu, suffix)
    201 num_rows = 0
    202 for part in subgraph.parts:
--> 203     table = subgraph[part]
    204     writer.add_data(table)
    205     num_rows += len(table)

File /nvtabular/nvtabular/io/dask.py:54, in DaskSubgraph.__getitem__(self, part)
     52 key = (self.name, part)
     53 dsk = self.subgraph.cull({key})
---> 54 return dask.get(dsk, key)

File /usr/local/lib/python3.8/dist-packages/dask/local.py:553, in get_sync(dsk, keys, **kwargs)
    548 """A naive synchronous version of get_async
    549 
    550 Can be useful for debugging.
    551 """
    552 kwargs.pop("num_workers", None)  # if num_workers present, remove it
--> 553 return get_async(
    554     synchronous_executor.submit,
    555     synchronous_executor._max_workers,
    556     dsk,
    557     keys,
    558     **kwargs,
    559 )

File /usr/local/lib/python3.8/dist-packages/dask/local.py:496, in get_async(submit, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, chunksize, **kwargs)
    494 while state["waiting"] or state["ready"] or state["running"]:
    495     fire_tasks(chunksize)
--> 496     for key, res_info, failed in queue_get(queue).result():
    497         if failed:
    498             exc, tb = loads(res_info)

File /usr/lib/python3.8/concurrent/futures/_base.py:437, in Future.result(self, timeout)
    435     raise CancelledError()
    436 elif self._state == FINISHED:
--> 437     return self.__get_result()
    439 self._condition.wait(timeout)
    441 if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:

File /usr/lib/python3.8/concurrent/futures/_base.py:389, in Future.__get_result(self)
    387 if self._exception:
    388     try:
--> 389         raise self._exception
    390     finally:
    391         # Break a reference cycle with the exception in self._exception
    392         self = None

File /usr/local/lib/python3.8/dist-packages/dask/local.py:538, in SynchronousExecutor.submit(self, fn, *args, **kwargs)
    536 fut = Future()
    537 try:
--> 538     fut.set_result(fn(*args, **kwargs))
    539 except BaseException as e:
    540     fut.set_exception(e)

File /usr/local/lib/python3.8/dist-packages/dask/local.py:234, in batch_execute_tasks(it)
    230 def batch_execute_tasks(it):
    231     """
    232     Batch computing of multiple tasks with `execute_task`
    233     """
--> 234     return [execute_task(*a) for a in it]

File /usr/local/lib/python3.8/dist-packages/dask/local.py:234, in <listcomp>(.0)
    230 def batch_execute_tasks(it):
    231     """
    232     Batch computing of multiple tasks with `execute_task`
    233     """
--> 234     return [execute_task(*a) for a in it]

File /usr/local/lib/python3.8/dist-packages/dask/local.py:225, in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    223     failed = False
    224 except BaseException as e:
--> 225     result = pack_exception(e, dumps)
    226     failed = True
    227 return key, result, failed

File /usr/local/lib/python3.8/dist-packages/dask/local.py:220, in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    218 try:
    219     task, data = loads(task_info)
--> 220     result = _execute_task(task, data)
    221     id = get_id()
    222     result = dumps((result, id))

File /usr/local/lib/python3.8/dist-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File /usr/local/lib/python3.8/dist-packages/dask/optimization.py:969, in SubgraphCallable.__call__(self, *args)
    967 if not len(args) == len(self.inkeys):
    968     raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
--> 969 return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))

File /usr/local/lib/python3.8/dist-packages/dask/core.py:149, in get(dsk, out, cache)
    147 for key in toposort(dsk):
    148     task = dsk[key]
--> 149     result = _execute_task(task, cache)
    150     cache[key] = result
    151 result = _execute_task(out, cache)

File /usr/local/lib/python3.8/dist-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File /usr/local/lib/python3.8/dist-packages/dask/utils.py:37, in apply(func, args, kwargs)
     35 def apply(func, args, kwargs=None):
     36     if kwargs:
---> 37         return func(*args, **kwargs)
     38     else:
     39         return func(*args)

File /nvtabular/nvtabular/workflow/workflow.py:437, in _transform_partition(root_df, workflow_nodes, additional_columns, capture_dtypes)
    435 for parent in node.parents_with_dependencies:
    436     parent_output_cols = _get_unique(parent.output_schema.column_names)
--> 437     parent_df = _transform_partition(root_df, [parent], capture_dtypes=capture_dtypes)
    438     if input_df is None or not len(input_df):
    439         input_df = parent_df[parent_output_cols]

File /nvtabular/nvtabular/workflow/workflow.py:485, in _transform_partition(root_df, workflow_nodes, additional_columns, capture_dtypes)
    483         elif len(output_df):
    484             if output_col_schema.dtype != output_df_schema.dtype:
--> 485                 raise TypeError(
    486                     f"Improperly matched output dtypes detected in {col_name},"
    487                     f" {output_col_schema.dtype} and {output_df_schema.dtype}"
    488                 )
    489 except Exception:
    490     LOG.exception("Failed to transform operator %s", node.op)

TypeError: Improperly matched output dtypes detected in c_156, float64 and int64
```

**Steps/Code to reproduce bug**
I cannot provide dataset here, and the issue does not occur on small dataset. 

**Environment details (please complete the following information):**
 - Environment location: [Bare-metal, Docker, Cloud(specify cloud provider)]
 - Method of NVTabular install: [conda, Docker, or from source]
   - If method of install is [Docker], provide `docker pull` & `docker run` commands used

`nvcr.io/nvidia/merlin/merlin-tensorflow-training:22.02` with the latest NVTabular main branch.

**Comment by benfred** [comment]
@rnyak is this fixed by #1398?

**Comment by benfred** [comment]
@karlhigley @rnyak is this still a problem?

**Comment by karlhigley** [comment]
Closing on the assumption that this was resolved by #1398

**Comment by rnyak** [comment]
@karlhigley and @benfred this is not a problem for this use-case but I was getting similar error when using JoinExternal op that I shared in the slack thread.
